app/feature/feature-chat/src/main/kotlin/com/hedvig/android/feature/chat/ChatViewModelNew.kt

```diff
@@ -0,0 +1,161 @@
+import androidx.compose.runtime.setValue
+import androidx.lifecycle.viewModelScope
+import com.hedvig.android.core.demomode.Provider
+import com.hedvig.android.feature.chat.data.ChatMessage
+import com.hedvig.android.hanalytics.featureflags.FeatureManager
+import com.hedvig.android.molecule.android.MoleculeViewModel
+import com.hedvig.android.molecule.public.MoleculePresenter
+import com.hedvig.android.molecule.public.MoleculePresenterScope
+import java.io.File
+import kotlin.time.Duration.Companion.seconds
+import kotlinx.collections.immutable.PersistentList
+import kotlinx.collections.immutable.persistentListOf
+import kotlinx.collections.immutable.toPersistentList
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.datetime.Instant
+
+internal class ChatViewModelNew(
+  chatRepository: Provider<ChatRepositoryNew>,
+  chatClosedTracker: ChatEventStore,
+  featureManager: FeatureManager,
+) : MoleculeViewModel<ChatEventNew, ChatUiState>(
+  ChatUiState(),
+  ChatPresenter(
+    chatRepository = chatRepository,
+    chatClosedTracker = chatClosedTracker,
+    featureManager = featureManager,
+  ),
+) {
+  init {
+    emit(ChatEventNew.FetchMessages())
+    viewModelScope.launch {
+      while (true) {
+        delay(5.seconds)
+        emit(ChatEventNew.PollMessages)
+      }
+    }
+  }
+}
+
+internal class ChatPresenter(
+  private val chatRepository: Provider<ChatRepositoryNew>,
+  private val chatClosedTracker: ChatEventStore,
+  private val featureManager: FeatureManager,
+) : MoleculePresenter<ChatEventNew, ChatUiState> {
+
+  @Composable
+  override fun MoleculePresenterScope<ChatEventNew>.present(lastState: ChatUiState): ChatUiState {
+    var isLoadingChat by remember { mutableStateOf(lastState.isLoadingChat) }
+    var isLoadingNewMessages by remember { mutableStateOf(lastState.isLoadingNewMessages) }
+    var messages by remember { mutableStateOf(lastState.messages) }
+    var isSendingMessage by remember { mutableStateOf(lastState.isSendingMessage) }
+    var canFetchMoreMessages by remember { mutableStateOf(lastState.canFetchMoreMessages) }
+    var canFetchUntil by remember { mutableStateOf(lastState.canFetchUntil) }
+    var errorMessage by remember { mutableStateOf(lastState.errorMessage) }
+
+    var fetchId by remember { mutableIntStateOf(0) }
+    var fetchUntil by remember { mutableStateOf<Instant?>(null) }
+    var fileState by remember { mutableStateOf<File?>(null) }
+    var messageInputState by remember { mutableStateOf<String?>(null) }
+
+    CollectEvents { event ->
+      when (event) {
+        is ChatEventNew.FetchMessages -> fetchUntil = event.until
+        ChatEventNew.PollMessages -> fetchId++
+        is ChatEventNew.SendFileMessage -> fileState = event.file
+        is ChatEventNew.SendTextMessage -> messageInputState = event.message
+        is ChatEventNew.DismissError -> errorMessage = null
+      }
+    }
+
+    LaunchedEffect(fetchId, fetchUntil) {
```

Comment:
Adding `fetchUntil` will re-trigger fetching new messages. It seems like to me when this call succeeds, fetchUntil always changes, so we'll basically always be doing 2 fetches here back to back, is this what we want? I suppose the 3rd time won't happen then, since the 2nd fetch will fetch the exact same fetchUntil value (old fetchUntil will be == the new fetchUntil) so this will not re-trigger.